### PR TITLE
Make Metrics consistent with Go client.

### LINKED ISSUE #102: Make tracer metrics consistent with Go client
https://github.com/jaegertracing/jaeger/issues/572

### PR DESCRIPTION
**Pull request by eundoosong**
Make Metrics consistent with Go client.

Now the metric name shows consistent with go-client as you see below prometheus test I did.
- in sampler, add SamplerMetrics which has 4 types of metrics, as go-client has.
- in tracer, add spans_finished 

Resolves #102 

Please review.

```
# HELP jaeger:traces jaeger:traces
# TYPE jaeger:traces counter
jaeger:traces{sampled="n",state="started"} 6.0
jaeger:traces{sampled="y",state="started"} 0.0
jaeger:traces{sampled="y",state="joined"} 0.0
jaeger:traces{sampled="n",state="joined"} 0.0
# HELP jaeger:sampler_updates jaeger:sampler_updates
# TYPE jaeger:sampler_updates counter
jaeger:sampler_updates{result="err"} 0.0
jaeger:sampler_updates{result="ok"} 0.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="2",minor="7",patchlevel="10",version="2.7.10"} 1.0
# HELP jaeger:sampler_queries jaeger:sampler_queries
# TYPE jaeger:sampler_queries counter
jaeger:sampler_queries{result="err"} 0.0
jaeger:sampler_queries{result="ok"} 0.0
# HELP jaeger:started_spans jaeger:started_spans
# TYPE jaeger:started_spans counter
jaeger:started_spans{sampled="n"} 12.0
jaeger:started_spans{sampled="y"} 0.0
# HELP jaeger:finished_spans jaeger:finished_spans
# TYPE jaeger:finished_spans counter
jaeger:finished_spans 0.0
# HELP jaeger:reporter_spans jaeger:reporter_spans
# TYPE jaeger:reporter_spans counter
jaeger:reporter_spans{result="dropped"} 0.0
jaeger:reporter_spans{result="socket_error"} 0.0
jaeger:reporter_spans{result="err"} 0.0
jaeger:reporter_spans{result="ok"} 0.0
```

Signed-off-by: eundoo-song <eundoo.song@gmail.com>

### CODE REVIEW

**Comment by eundoosong** [comment]
It looks like the uncovered line(return PeriodicCallback) is not related to this fix. 
Anyone knows how to fix this?

**Comment by yurishkuro** [comment]
@eundoosong Sweet!  Thanks for contributing. Don't worry about the coverage, it looks like a glitch.

@jpkrohling could you take a look, since you just did the same in Java?

**Comment by jpkrohling** [comment]
LGTM. Looks like the Python version is missing a gauge (`reporter_queue_length`) and has an extra counter (`jaeger:reporter_spans{'result': 'socket_error}`), but I think those differences are OK.

**Comment by eundoosong** [comment]
@jpkrohling Thanks for your review. 
I think we can add reporter_queue_length gauge just after span batch submission like this.
Tornado provides an api to know current size of queue.
http://www.tornadoweb.org/en/stable/_modules/tornado/queues.html#Queue.qsize
```python
if spans:
    yield self._submit(spans)
    self.metrics.reporter_queue_length(self.queue.qsize())
    for _ in spans:
        self.queue.task_done()
    spans = spans[:0]
```
If it's ok, I'll update.

**Comment by jpkrohling** [comment]
I'm not familiar with Tornado (or Python in general, for that matter), but looks OK to me.

**Comment by eundoosong** [comment]
@yurishkuro  Is there any reason reporter_queue_length gauge is missed in python? Do you think it should be added as above?

**Comment by yurishkuro** [comment]
@eundoosong yes, please add it. Only I would add it after that if statement, i.e. as the last statement of the while loop (or alternatively the first statement).

Re `jaeger:reporter_spans{'result': 'socket_error}` counter, I am not sure it's worth having it separate from the `err` counter, as it makes aggregations less obvious.

**Comment by eundoosong** [comment]
@yurishkuro  @jpkrohling  reporter_socket(removed) is replaced by reporter_failure and reporter_queue_length is moved.  
Is this ready to be merged, or is there anything else missing?

File: jaeger_client/sampler.py [hunk 1]
```diff
@@ -435,14 +435,15 @@ def _update_sampler(self, response):
...
-                self.sampler_errors(1)
+                self.metrics.sampler_update_failure(1)
                 self.error_reporter.error(
                     'Fail to update sampler'
                     'from jaeger-agent: %s [%s]', e, response)
 
     def _update_adaptive_sampler(self, per_operation_strategies):
         if isinstance(self.sampler, AdaptiveSampler):
             self.sampler.update(per_operation_strategies)
+            self.metrics.sampler_updated(1)
```
Comment (yurishkuro) [inline jaeger_client/sampler.py]: this should always run, for `else:` part as well

**Comment by eundoosong** [comment]
@yurishkuro  moved sampler_updated metric out of the condition

**Comment by yurishkuro** [comment]
@eundoosong have you tried this in a real service with a real metrics library?

**Comment by eundoosong** [comment]
@yurishkuro  I tried this using my prometheus client as you see my first comment above 
code: https://github.com/eundoosong/jaeger-client-python/commit/78aa262ef5d20c8a7b8789b25591584215680088#diff-08bb54ec260941e2becb5054785b7fa2
It was for test, but I'm planning to use this in a real service with Prometheus and Grafana.

**Comment by yurishkuro** [comment]
Thanks, @eundoosong!

Do you think you can add a PR for `PrometheusMetricsFactory`? It think it would be great to have it in Jaeger client. The only gotcha is that we don't want to have a runtime dependency on prometheus client, so I would design the PrometheusMetricsFactory to import prom client conditionally (blow up if not found).

**Comment by eundoosong** [comment]
@yurishkuro Sure, I can add PrometheusMetricsFactory. PR will be created after tidying code and a try to remove a runtime dependency on prometheus client.
And I'd love to contribute the same in nodejs since our another main language for tracing client is nodejs.
related issue: https://github.com/jaegertracing/jaeger-client-node/issues/178

**Pull request merged**